plugins/ltac/tacintern.ml

```diff
@@ -95,9 +95,16 @@ let intern_string_or_var = intern_or_var (fun (s : string) -> s)
 let intern_global_reference ist qid =
   if qualid_is_ident qid && find_var (qualid_basename qid) ist then
     ArgVar (make ?loc:qid.CAst.loc @@ qualid_basename qid)
-  else
-    try ArgArg (qid.CAst.loc,locate_global_with_alias qid)
-    with Not_found -> Nametab.error_global_not_found qid
+  else if not !strict_check && qualid_is_ident qid && find_hyp (qualid_basename qid) ist then
+    let id = qualid_basename qid in
+    ArgArg (qid.CAst.loc, GlobRef.VarRef id)
+  else match locate_global_with_alias qid with
+  | r -> ArgArg (qid.CAst.loc, r)
+  | exception Not_found ->
+    if not !strict_check && qualid_is_ident qid then
+      let id = qualid_basename qid in
+      ArgArg (qid.CAst.loc, GlobRef.VarRef id)
+    else Nametab.error_global_not_found qid
```

Comment:
Do we want this before or after the attempt to interpret statically? I mean, what do we want here (in non-strict mode):
```
Tactic Notation "foo" reference(n) := idtac n.

Goal forall n : nat, n = 0.
Proof.
intros nat; foo nat. (* the type or the variable: the PR says the type *)
Undo.
intros nat. foo nat. (* the type or the variable: the PR says the variable *)
```

Reply:
I wrote it that way to minimize breakage in #12256. I think the current behaviour (after the removal of the strict check in the second branch) is more natural.